### PR TITLE
Support type inheritance

### PR DESCRIPTION
**Pull request by Vulthil**
This pull request should support type inheritance.

You might find something which I have forgot to check for, or maybe you know a better way to compare and make sure the base type is included.

### CODE REVIEW

File: src/Tapper/TypeScriptCodeGenerator.cs [hunk 1]
```diff
@@ -104,6 +104,17 @@ private void AddClassOrStruct(INamedTypeSymbol typeSymbol, ref CodeWriter writer
         }
 
         writer.Append('}');
+
+        if (typeSymbol.BaseType != null &&
+            typeSymbol.BaseType.IsType &&
+            typeSymbol.BaseType.MetadataName != typeof(object).Name)
```
Comment (nenoNaninu) [inline src/Tapper/TypeScriptCodeGenerator.cs:110]:
```suggestion
        if (typeSymbol.BaseType is not null &&
            typeSymbol.BaseType.IsType &&
            typeSymbol.BaseType.SpecialType != SpecialType.System_Object)
```

**Comment by nenoNaninu** [comment]
Thanks for the pull request!

This code lacks consideration for using types in other namespaces as basic types. For example, the following C# code cannot be correctly transpiled to TypeScript.

```cs
namespace Space1
{
    [TranspilationSource]
    public class CustomType1
    {
        public float Value;
        public DateTime Name;
    }
}

namespace Space2
{
    using Space1;

    [TranspilationSource]
    public class CustomType2 : CustomType1
    {
        public float Value2;
        public DateTime Name2;
    }
}
```

**Comment by Vulthil** [comment]
Yes, I was thinking about that.

How are you handling the imports in the other parts of the transpilation process, and could the same be used in this case ?

**Comment by nenoNaninu** [comment]
It is processed in the following code.

https://github.com/nenoNaninu/Tapper/blob/main/src/Tapper/TypeScriptCodeGenerator.cs#L40-L52

I think that the purpose can be achieved if the base type is included in this process as well as the member type.

**Comment by Vulthil** [comment]
Added another HeaderTest to test for correct import of a type from a different namespace.

File: src/Tapper/TypeScriptCodeGenerator.cs [hunk 2]
```diff
@@ -37,12 +37,16 @@ public void AddHeader(IGrouping<INamespaceSymbol, INamedTypeSymbol> types, ref C
         writer.Append($"/* eslint-disable */{_newLine}");
         writer.Append($"/* tslint:disable */{_newLine}");
 
-        var diffrentNamespaceTypes = types
-            .SelectMany(static x => x.GetPublicFieldsAndProperties().IgnoreStatic())
+        var memberTypes = types.SelectMany(static x => x.GetPublicFieldsAndProperties().IgnoreStatic());
```
Comment (nenoNaninu) [inline src/Tapper/TypeScriptCodeGenerator.cs:40]:
```suggestion
        var memberTypes = types
            .SelectMany(static x => x.GetPublicFieldsAndProperties().IgnoreStatic())
            .SelectMany(RoslynExtensions.GetRelevantTypesFromMemberSymbol);
```

File: src/Tapper/TypeScriptCodeGenerator.cs [hunk 3]
```diff
@@ -37,12 +37,16 @@ public void AddHeader(IGrouping<INamespaceSymbol, INamedTypeSymbol> types, ref C
...
+        var baseTypes = types.Where(static x => x.BaseType is not null && x.BaseType.IsType && !x.BaseType.SpecialType.Equals(SpecialType.System_Object)).Select(static x => x.BaseType!);
```
Comment (nenoNaninu) [inline src/Tapper/TypeScriptCodeGenerator.cs:41]:
```suggestion
        var baseTypes = types
            .Where(static x => x.BaseType is not null
                && x.BaseType.IsType
                && x.BaseType.SpecialType != SpecialType.System_Object)
            .Select(static x => x.BaseType!);
```

File: src/Tapper/TypeScriptCodeGenerator.cs [hunk 4]
```diff
@@ -37,12 +37,16 @@ public void AddHeader(IGrouping<INamespaceSymbol, INamedTypeSymbol> types, ref C
...
+
+        var diffrentNamespaceTypes = memberTypes
             .SelectMany(RoslynExtensions.GetRelevantTypesFromMemberSymbol)
+            .Concat(baseTypes)
```
Comment (nenoNaninu) [inline src/Tapper/TypeScriptCodeGenerator.cs:45]:
```suggestion
        var diffrentNamespaceTypes = memberTypes
            .Concat(baseTypes)
```

File: src/Tapper/TypeScriptCodeGenerator.cs [hunk 5]
```diff
@@ -37,12 +37,16 @@ public void AddHeader(IGrouping<INamespaceSymbol, INamedTypeSymbol> types, ref C
...
             .OfType<INamedTypeSymbol>()
             .Where(x => !SymbolEqualityComparer.Default.Equals(x.ContainingNamespace, types.Key)
-                && _sourceTypes.Contains(x, SymbolEqualityComparer.Default))
+                && _sourceTypes.Contains(x, SymbolEqualityComparer.Default)
+            )
```
Comment (nenoNaninu) [inline src/Tapper/TypeScriptCodeGenerator.cs:49]:
```suggestion
                && _sourceTypes.Contains(x, SymbolEqualityComparer.Default))
```

**Review by nenoNaninu — APPROVED** [review]
Thanks for your contribution!

**Comment by serard-bimandco** [comment]
Hello, has it been released ?
Thank you !

**Comment by nenoNaninu** [comment]
Yes, it has already been released. Please try it.

**Comment by serard-bimandco** [comment]
Thank you, saw it, was my fault not using correctly, missing some [] annotations.

**Comment by nenoNaninu** [comment]
Please post a small example.

**Pull request merged**